File: appwin_attribution/android/src/main/kotlin/io/appwin/flutter/attribution/AppwinAttributionPlugin.kt
package io.appwin.flutter.attribution

import android.os.Build
import io.appwin.attribution.AdvertisingConsent
import io.appwin.attribution.AppwinAttribution
import io.appwin.core.availability.AppwinInitResult
import io.appwin.core.availability.AppwinInitStatus
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Dart-to-Kotlin glue for the Attribution product.
 *
 * Mirrors `AppwinAttributionPlugin.swift`: same method names, same arguments,
 * same result shapes. `requestTrackingAuthorization` resolves `true` here -
 * Android has no ATT, tracking is not OS-gated.
 */
class AppwinAttributionPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {

  private var channel: MethodChannel? = null
  private var scope: CoroutineScope? = null

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    channel = MethodChannel(binding.binaryMessenger, "appwin_attribution").also {
      it.setMethodCallHandler(this)
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel?.setMethodCallHandler(null)
    channel = null
    scope?.cancel()
    scope = null
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "getPlatformVersion" -> result.success("Android ${Build.VERSION.RELEASE}")
      "initialize" -> {
        val scope = scope ?: return result.error("not_attached", "plugin detached", null)
        scope.launch {
          result.success(encode(AppwinAttribution.initialize()))
        }
      }
      "setAdvertisingConsent" -> {
        AppwinAttribution.setAdvertisingConsent(parseConsent(call.argument<String>("consent")))
        result.success(null)
      }
      "requestTrackingAuthorization" -> result.success(true)
      "setAdSignalsDebugMode" -> {
        AppwinAttribution.setAdSignalsDebugMode(call.argument<Boolean>("enabled") ?: false)
        result.success(null)
      }
      else -> result.notImplemented()
    }
  }

  private fun parseConsent(raw: String?): AdvertisingConsent = when (raw) {
    "granted" -> AdvertisingConsent.GRANTED
    "denied" -> AdvertisingConsent.DENIED
    else -> AdvertisingConsent.UNKNOWN
  }

  /** Same shape as the iOS plugin: status + optional reason. */
  private fun encode(result: AppwinInitResult): Map<String, Any?> = when (result.status) {
    AppwinInitStatus.READY -> mapOf("status" to "ready")
    AppwinInitStatus.NOT_CONFIGURED -> mapOf("status" to "notConfigured")
    AppwinInitStatus.UNKNOWN -> mapOf("status" to "unknown")
    AppwinInitStatus.UNAVAILABLE ->
      mapOf("status" to "unavailable", "reason" to result.reason?.key)
  }
}

File: appwin_attribution/ios/appwin_attribution/Sources/appwin_attribution/AppwinAttributionPlugin.swift
import Flutter
import UIKit
import AppwinCore
import AppwinAttribution

// Native modules are declared as `s.dependency` in the podspec (ADR-0020,
// Firebase/FlutterFire pattern). This file is the only glue; the
// implementations live in AppwinCore and AppwinAttribution.
public class AppwinAttributionPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "appwin_attribution", binaryMessenger: registrar.messenger())
        let instance = AppwinAttributionPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    @MainActor
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        case "initialize":
            // Availability, not configuration: `AppwinCore.configure` is the
            // host app's job and runs through the appwin_core plugin.
            Task {
                let verdict = await AppwinAttribution.initialize()
                await MainActor.run { result(Self.encode(verdict)) }
            }
        case "setAdvertisingConsent":
            let raw = (call.arguments as? [String: Any])?["consent"] as? String
            AppwinAttribution.setAdvertisingConsent(Self.parseConsent(raw))
            result(nil)
        case "requestTrackingAuthorization":
            Task {
                let granted = await AppwinAttribution.requestTrackingAuthorization()
                await MainActor.run { result(granted) }
            }
        case "setAdSignalsDebugMode":
            let enabled = (call.arguments as? [String: Any])?["enabled"] as? Bool ?? false
            AppwinAttribution.setAdSignalsDebugMode(enabled)
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private static func parseConsent(_ raw: String?) -> AdvertisingConsent {
        switch raw {
        case "granted": return .granted
        case "denied": return .denied
        default: return .unknown
        }
    }

    /// Maps the native result onto what the Dart side parses. Same shape
    /// as every other product plugin: status + optional reason.
    private static func encode(_ verdict: AppwinInitResult) -> [String: Any] {
        switch verdict {
        case .ready:
            return ["status": "ready"]
        case .notConfigured:
            return ["status": "notConfigured"]
        case .unknown:
            return ["status": "unknown"]
        case .unavailable(let reason):
            return ["status": "unavailable", "reason": reason.rawValue]
        }
    }
}
